### PR TITLE
[SPARK-30104][SQL][FOLLOWUP] V2 catalog named 'global_temp' should always be masked

### PR DESCRIPTION
**Pull request by imback82**
### What changes were proposed in this pull request?

This is a follow up to #26741 to address the following:
1. V2 catalog named `global_temp` should always be masked.
2. #26741 introduces `CatalogAndIdentifer` that supersedes `CatalogObjectIdentfier`. This PR removes `CatalogObjectIdentfier` and its usages and replace them with `CatalogAndIdentifer`.
3. `CatalogObjectIdentifier(catalog, ident) if !isSessionCatalog(catalog)` and `CatalogObjectIdentifier(catalog, ident) if isSessionCatalog(catalog)` are replaced with `NonSessionCatalogAndIdentifier` and `SessionCatalogAndIdentifier` respectively.

### Why are the changes needed?

To fix an existing with handling v2 catalog named `global_temp` and to simplify the code base.

### Does this PR introduce any user-facing change?

No

### How was this patch tested?

Added new tests.

### CODE REVIEW

File: sql/catalyst/src/test/scala/org/apache/spark/sql/connector/catalog/LookupCatalogSuite.scala [hunk 1]
```diff
@@ -46,13 +49,16 @@ class LookupCatalogSuite extends SparkFunSuite with LookupCatalog with Inside {
...
     manager
   }
 
-  test("catalog object identifier") {
+  test("catalog and identifier") {
     Seq(
       ("tbl", sessionCatalog, Seq.empty, "tbl"),
       ("db.tbl", sessionCatalog, Seq("db"), "tbl"),
+      (s"$globalTempDB.tbl", sessionCatalog, Seq(globalTempDB), "tbl"),
+      (s"$globalTempDB.ns1.ns2.tbl", catalogs(globalTempDB), Seq("ns1", "ns2"), "tbl"),
```
Comment (imback82) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/connector/catalog/LookupCatalogSuite.scala]: @cloud-fan just to confirm, `global_temp` can be resolved to a v2 catalog named `global_temp`, right?

Comment (cloud-fan) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/connector/catalog/LookupCatalogSuite.scala]: no we can't. `global_temp` is kind of a reserved catalog that only contains global temp views. If a user-registered catalog is also named `global_temp`, it should be masked.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 2]
```diff
@@ -357,11 +357,11 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
     val canUseV2 = lookupV2Provider().isDefined
 
     session.sessionState.sqlParser.parseMultipartIdentifier(tableName) match {
-      case CatalogObjectIdentifier(catalog, ident) if !isSessionCatalog(catalog) =>
+      case NonSessionCatalogAndIdentifier(catalog, ident) =>
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: This looks much better (BTW, this is not a replacement from `CatalogObjectIdentifier` to `CatalogAndIdentifer`)

Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: I like this change. Could you update the PR description more precisely?

Comment (imback82) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: Updated. Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/LookupCatalog.scala [hunk 3]
```diff
@@ -117,7 +104,7 @@ private[sql] trait LookupCatalog extends Logging {
       assert(nameParts.nonEmpty)
       if (nameParts.length == 1) {
         Some((currentCatalog, Identifier.of(Array(), nameParts.head)))
-      } else if (nameParts.length == 2 && nameParts.head.equalsIgnoreCase(globalTempDB)) {
+      } else if (nameParts.head.equalsIgnoreCase(globalTempDB)) {
```
Comment (imback82) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/LookupCatalog.scala]: Now that we handle the single-name part differently, we don't need the check here any more. This case will now resolve to the session catalog (mask v2 catalog named `global_temp`), and analysis will fail downstream if the name has more than two parts.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**